### PR TITLE
Prioritized experience replay

### LINKED ISSUE #1242: Prioritized Experience Replay for DQN
### 🚀 Feature

Prioritized Experience Replay for DQN

### Motivation

_No response_

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo

### PR DESCRIPTION
**Pull request by AlexPasqua**
## Description

Implementation of prioritized replay buffer for DQN.
Closes #1242 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

In accordance with #1242

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by AlexPasqua** [comment]
@araffin could you (or anyone) please have a look at the 2 pytype errors? I don't quite understand how to fix them

File: stable_baselines3/dqn/prioritized_replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        weights = weights / weights.max()
+
+        env_indices = np.random.randint(0, high=self.n_envs, size=(len(sample_idxs),))
+
+        if self.optimize_memory_usage:
+            next_obs = self._normalize_obs(
+                self.observations[(np.array(sample_idxs) + 1) % self.buffer_size, env_indices, :], env
+            )
+        else:
+            next_obs = self._normalize_obs(self.next_observations[sample_idxs, env_indices, :], env)
```
Comment (araffin) [inline stable_baselines3/dqn/prioritized_replay_buffer.py:212]: pytype thinks this value might None, which is apparently true (see `buffers.py` for solutions)

File: stable_baselines3/dqn/prioritized_replay_buffer.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+        batch = (
+            self._normalize_obs(self.observations[sample_idxs, env_indices, :], env),
+            self.actions[sample_idxs, env_indices, :],
+            next_obs,
+            self.dones[sample_idxs],
+            self.rewards[sample_idxs],
+        )
+        return ReplayBufferSamples(*tuple(map(self.to_torch, batch)))
```
Comment (araffin) [inline stable_baselines3/dqn/prioritized_replay_buffer.py:221]: you can probably ignore pytype here

File: stable_baselines3/dqn/prioritized_replay_buffer.py (hunk 1 above)
Comment (AlexPasqua) [reply stable_baselines3/dqn/prioritized_replay_buffer.py:212]: `self.next_observations` might be None if `self.optimize_memory_usage` is True, but then we wouldn't enter the `else` at line 212.
The same thing happens in `buffers.py`, so I don't quite understand where the difference is

Comment (AlexPasqua) [reply stable_baselines3/dqn/prioritized_replay_buffer.py:212]: Hello @araffin,
could you have a look at this (or the PR in general) when you have time?

After that I'll work on the tests, documentation etc. in order to complete the PR.
Cheers!

Comment (araffin) [reply stable_baselines3/dqn/prioritized_replay_buffer.py:212]: I'm pretty busy right now (writing paper for end of the month) and I was waiting for the tests to pass, but I'll try to have a look beginning of October ;)

**Comment by AlexPasqua** [comment]
Thanks @araffin !
Out of curiosity, may I ask why the switch between torch and numpy for the backend?

**Comment by araffin** [comment]
> Thanks @araffin ! Out of curiosity, may I ask why the switch between torch and numpy for the backend?

to be consistent with the rest of the buffers and because PyTorch is not needed here (no gpu computation needed).

**Comment by AlexPasqua** [comment]
Hello @araffin ,
as you moved the code to "common", I suppose you plan to make it usable in algorithms other than DQN. At this point, wouldn't it be clearer to put the code into `common/buffers.py`? Let me know, in case, I will move it there.

**Comment by araffin** [comment]
> At this point, wouldn't it be clearer to put the code into common/buffers.py?

yes probably, but the most important thing for now is to test the implementation (performance test, check we can reproduce the results from the paper), document it and add additional tests/doc (for sumtree for instance).

**Comment by araffin** [comment]
> performance test, check we can reproduce the results from the paper

After some initial test on Breakout following hyperparameters from the paper, the run didn't improve or worsen DQN performance so far...
I will try on other envs (it would be nice if you could help).

**Comment by AlexPasqua** [comment]
> After some initial test on Breakout following hyperparameters from the paper, the run didn't improve or worsen DQN performance so far... I will try on other envs (it would be nice if you could help).

Thanks for starting to test it!
These days I'm travelling, and also writing a paper after work, but I'll try to squeeze some tests in

**Comment by AlexPasqua** [comment]
I've also done some initial tests and it looks like PER might lead to a slightly faster convergence, for example on cartpole, but nothing super evident unfortunately.
Next I'd like to properly reproduce some of the paper's experiment, but computational power could become a bit of an issue for me

File: stable_baselines3/dqn/dqn.py [hunk 3]
```diff
@@ -208,8 +209,20 @@ def train(self, gradient_steps: int, batch_size: int = 100) -> None:
...
+                # Weighted Huber loss using importance sampling weights
+                loss = (replay_data.weights * th.where(td_error < 1.0, 0.5 * td_error**2, td_error - 0.5)).mean()
+                # Update priorities, they will be proportional to the td error
+                assert replay_data.leaf_nodes_indices is not None, "Node leaf node indices provided"
+                self.replay_buffer.update_priorities(
+                    replay_data.leaf_nodes_indices, td_error, self._current_progress_remaining
+                )
+            else:
+                # Compute Huber loss (less sensitive to outliers)
+                loss = F.smooth_l1_loss(current_q_values, target_q_values)
```
Comment (qgallouedec) [inline stable_baselines3/dqn/dqn.py:225]: @AlexPasqua Ideally, we'd like to be able to associate it with all off-policy algo's without adaptation, but I don't see a simple way of doing it at this stage.
Also related, we had discussed not modifying DQN: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/127#issuecomment-1381530029

**Comment by richardjozsa** [comment]
Just a comment, I've tested this implementation with QR-DQN with Vecenv multiple environment but it fails because of the missing part.

But good job to start the work on it! I hope it will be merged soon! 👍

File: stable_baselines3/dqn/dqn.py (hunk 3 above)
Comment (jbial) [reply stable_baselines3/dqn/dqn.py:225]: I'm interested in this PR. Since every algo-specific `train` method includes a `replay_buffer.sample` line, couldn't we just additionally add a `replay_buffer.update` line? The update function could take in the current and target q values whenever a value function is present or maybe even all the local variables. It would do nothing for the vanilla replay buffer. Would this be an acceptable modification?

Comment (qgallouedec) [reply stable_baselines3/dqn/dqn.py:225]: Thanks for your comment!
How do you handle the loss in your proposal?

Comment (jbial) [reply stable_baselines3/dqn/dqn.py:225]: If we want this to work for general off-policy algorithms, we could update the ReplayBufferSample-like classes to additionally include an `importance_sampling_weight` attribute which would be updated from the `replay_buffer.update` method.

Then I see two ways to handle the loss under this interface:

1. Estimate TD error from the loss as such:
```
losses = loss_fn(current_q_values, target_q_values, reduction='none')

# e.g. If loss is L2, then it's basically th.sqrt(loss). If loss is L1, td_error = loss
td_error = importance_sampling_weight * function_to_approx_td_error(losses)  

loss = losses.mean()
```
Obviously the downside of this is that it requires hand engineering for the different types of loss functions or priority metrics.

2. Make any value-based `train` methods "td-error" centric in the sense that we always compute `td_error = importance_sampling_weight * th.abs(current_q_values - target_q_values)` first, then the loss `loss = loss_fn(td_error)`. The downsides of this approach is that we cant use the pytorch api for computing the loss, and would have to write functions for those.

Either approach requires computing a `td_error` variable which unfortunately requires somewhat intrusive code changes. What do you think?

Comment (araffin) [reply stable_baselines3/dqn/dqn.py:225]: maybe to make things clearer: my plan is not to have for all algorithms, mainly for two reasons:

1. Keep the code concise (in fact, I would like to have RAINBOW and keep vanilla DQN, see https://github.com/DLR-RM/stable-baselines3/issues/622)
2. I don't think it works for entropy-RL algorithms (SAC and derivates), so it would be limited to DQN/QR-DQN and TD3

Comment (jbial) [reply stable_baselines3/dqn/dqn.py:225]: In that case, I'd be happy to contribute to the Rainbow DQN feature request. Dueling DQN and double DQN are straightforward to implement, and @araffin PER implementation seems to work. Also C51 is already implemented in contrib. We would just need n-step returns and noisy nets right?

Comment (araffin) [reply stable_baselines3/dqn/dqn.py:225]: "just" yes, I would be happy to receive such PR =)
the main thing is to benchmark the implementation and reproduce the published results.
This PR is also still open because I was not satisfied by the result of DQN + PER (I couldn't see significant different with respect to DQN).

Comment (araffin) [reply stable_baselines3/dqn/dqn.py:225]: One thing I had in mind was to implement CNN for SBX (https://github.com/araffin/sbx) in order to iterate faster and check the PER, but I had no time to do so until now...

Comment (jbial) [reply stable_baselines3/dqn/dqn.py:225]: Why don't we implement the toy environment from figure 1 of https://arxiv.org/pdf/1511.05952 as the PER benchmark? It would be a simpler initial check for correctness than the Atari environments

Comment (araffin) [reply stable_baselines3/dqn/dqn.py:225]: The toy environment can be a start for fast iteration and debugging, but what we learn in the past is that subtle bugs only show up when doing more complex task (see https://github.com/DLR-RM/stable-baselines3/issues/48 and https://github.com/DLR-RM/stable-baselines3/issues/47 where we found bugs like PyTorch and TF RMSProp are not the same)

Comment (jbial) [reply stable_baselines3/dqn/dqn.py:225]: I see, will definitely work towards it!

File: stable_baselines3/common/prioritized_replay_buffer.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        beta: float = 0.4,
+        final_beta: float = 1.0,
+        optimize_memory_usage: bool = False,
+        min_priority: float = 1e-8,
+    ):
+        super().__init__(buffer_size, observation_space, action_space, device, n_envs)
+
+        assert optimize_memory_usage is False, "PrioritizedReplayBuffer doesn't support optimize_memory_usage=True"
+
+        self.min_priority = 1e-8
```
Comment (janakact) [inline stable_baselines3/common/prioritized_replay_buffer.py:129]: Small mistake. Change this to `self.min_priority = min_priority` instead.  ;)

**Comment by jbial** [comment]
I've just tried validating the implementation on blind cliffwalk and it seems much slower (~an order of magnitude) than the uniform replay buffer. The results below are for a one seed:
Screenshot 2024-05-27 at 10 20 10 PM

Not sure why this is. The details for blind cliffwalk are a bit vague from the paper (no code available as well), but I've tried to implement it as close to the description as possible.

Code for the test is in this gist:
https://gist.github.com/jbial/105299c00dc3bb7960f0f17f2fc4d6c9

File: stable_baselines3/common/prioritized_replay_buffer.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        # probability of sampling transition i as P(i) = p_i^alpha / \sum_{k} p_k^alpha
+        # where p_i > 0 is the priority of transition i.
+        probs = priorities / self.tree.total_sum
+
+        # Importance sampling weights.
+        # All weights w_i were scaled so that max_i w_i = 1.
+        weights = (self.size() * probs) ** -self.beta
+        weights = weights / weights.max()
+
+        # TODO: add proper support for multi env
```
Comment (richardjozsa) [inline stable_baselines3/common/prioritized_replay_buffer.py:218]: How could we add proper support for multiple envs? Is there any idea? Does the random line below could work?

Comment (araffin) [reply stable_baselines3/common/prioritized_replay_buffer.py:218]: not sure yet, the random line below might work but we need to check if it won't affect performance first.

**Comment by araffin** [comment]
Some update from my part, I just added CNN support for SBX (SB3 + Jax) DQN, and it is 10x faster than the PyTorch equivalent: https://github.com/araffin/sbx/pull/49

That should allow to test and debug things more quickly on Atari (~1h40 for 10M steps instead of 15h =D)

Perf report: https://wandb.ai/openrlbenchmark/sbx?nw=nwuseraraffin (on-going)

File: stable_baselines3/common/prioritized_replay_buffer.py (hunk 5 above)
Comment (araffin) [reply stable_baselines3/common/prioritized_replay_buffer.py:218]: https://github.com/araffin/sbx/pull/50/commits/b5ce09105aa8df2701e85d30f0584380b2ee19f8 should be better, see https://github.com/araffin/sbx/pull/50

**Comment by araffin** [comment]
Some additional update: when trying to plug the PER implementation of this PR inside the Jax DQN implementation, the experience replay was the bottleneck (by a good margin, making things 40x slower...), so I investigated different ways to speed things up.

After playing with many different implementation (pure python, numpy, jax, jax jitted, ...), I decided to re-use the SB2 "SegmentTree" vectorized implementation and also implement proper multi-env support.
My current progress is here: https://github.com/araffin/sbx/pull/50

(still debugging, but at least I've got the first sign of life and this implementation is so much faster)

**Comment by richardjozsa** [comment]
Hey @araffin , it is great to hear that. Does SBX/Jax means this much speed improvement? 

If you think it is ready for testing I can give a try, just let me know when it is ready to be tested. :)

**Comment by araffin** [comment]
> Does SBX/Jax means this much speed improvement?

With the right parameters (see the exact command line argument for the RL Zoo in the OpenRL benchmark organization run on W&B), yes, around 10x faster.

> If you think it is ready for testing I can give a try, just let me know when it is ready to be tested. :)

SBX version is ready to be tested but so far, I didn't manage to see any gain from the PER. I also experienced some explosion in the qf value when using multiple env (so there is probably a bug here).
I'm also wondering if I need to implement double q-learning (easy) too to compare to the original paper.

**Comment by richardjozsa** [comment]
When I tested this PR I also noticed an explosion loss,  in that time I felt that it is because of the tweaking here and there. and I also noticed that it doesn't give me any advantage over a normal buffer(and I used Dobule DQN, even tried duelling), but I tried to tweak an N-step buffer which had a strong effect on the learning, AFAIK N-step(multi step) is also part of Rainbow and giving substantial part of the success.

The key parts are the distributional, PER and N-step parts, as far as I understand the concept. The others are kinda tasks specific parts and can be detrimental to use them.

**Comment by bilelsgh** [comment]
Ok, nice.
The PR for the propritized approzimation loss is pending, let me know if I can help for something.

**Comment by nf2bi89** [comment]
@araffin is the idea to move PER to sbx entirely as it already has multi-env implemented which this version doesn't? or is multi-env support not intended for this version? Also if the prioritized approximation loss implementation is similar enough to the original PER at a reduced computational cost doesn't this solve the exact speed issue you are currently seeking plus it already works with multi-envs? I have been unable to compare them yet though as it appears to ignore the passed buffer class and doesn't have a working example.
